src/main/scheduler/scheduler.c

```diff
@@ -350,79 +375,143 @@ static void readSchedulerLocals(task_t *selectedTask, uint8_t selectedTaskDynami
             }
             currentTimeUs = micros();
-            realtimeTaskRan = true;
 
-#if defined(USE_LATE_TASK_STATISTICS)
-           // Late, so make a note of the offending task
-            if (gyroTaskDelayUs < -1) {
-                if (lastTask) {
-                    lastTask->lateCount++;
+            lastTargetCycles = nextTargetCycles;
+
+#ifdef USE_GYRO_EXTI
+            gyroDev_t *gyro = gyroActiveDev();
+
+            // Bring the scheduler into lock with the gyro
+            if (gyro->gyroModeSPI != GYRO_EXTI_NO_INT) {
+                // Track the actual gyro rate over given number of cycle times and set the expected timebase
+                static uint32_t terminalGyroRateCount = 0;
+                static int32_t sampleRateStartCycles;
+
+                if ((terminalGyroRateCount == 0)) {
+                    terminalGyroRateCount = gyro->detectedEXTI + GYRO_RATE_COUNT;
+                    sampleRateStartCycles = nowCycles;
+                }
+
+                if (gyro->detectedEXTI >= terminalGyroRateCount) {
+                    // Calculate the number of clock cycles on average between gyro interrupts
+                    uint32_t sampleCycles = nowCycles - sampleRateStartCycles;
+                    desiredPeriodCycles = sampleCycles / GYRO_RATE_COUNT;
+                    sampleRateStartCycles = nowCycles;
+                    terminalGyroRateCount += GYRO_RATE_COUNT;
+                }
+
+                // Track the actual gyro rate over given number of cycle times and remove skew
+                static uint32_t terminalGyroLockCount = 0;
+                static int32_t accGyroSkew = 0;
+
+                int32_t gyroSkew = cmpTimeCycles(nextTargetCycles, gyro->gyroSyncEXTI) % desiredPeriodCycles;
+                if (gyroSkew > (desiredPeriodCycles / 2)) {
+                    gyroSkew -= desiredPeriodCycles;
+                }
+
+                accGyroSkew += gyroSkew;
+
+                if ((terminalGyroLockCount == 0)) {
+                    terminalGyroLockCount = gyro->detectedEXTI + GYRO_LOCK_COUNT;
+                }
+
+                if (gyro->detectedEXTI >= terminalGyroLockCount) {
+                    terminalGyroLockCount += GYRO_LOCK_COUNT;
+
+                    // Move the desired start time of the gyroTask
+                    lastTargetCycles -= (accGyroSkew/GYRO_LOCK_COUNT);
+                    DEBUG_SET(DEBUG_SCHEDULER_DETERMINISM, 3, clockCyclesTo10thMicros(accGyroSkew/GYRO_LOCK_COUNT));
+                    accGyroSkew = 0;
                 }
             }
-            lastTask = NULL;
 #endif
-        }
+       }
     }
 
-    if (!gyroEnabled || realtimeTaskRan || (gyroTaskDelayUs > GYRO_TASK_GUARD_INTERVAL_US)) {
-        // The task to be invoked
+    nowCycles = getCycleCounter();
+    schedLoopRemainingCycles = cmpTimeCycles(nextTargetCycles, nowCycles);
 
+    const timeUs_t currentTimeUs = micros();
```

Comment:
`currentTimeUs` may be unused if the expression on line 496 is false.  Perhaps move the call to `micros()` into the block after line 496 to avoid unnecessary calls to `micros()`

Reply:
Fixed in https://github.com/betaflight/betaflight/pull/10573/commits/39b9cac6729a73cedb8bec20452a5ff5dd361038